src/main/java/com/hubspot/imap/utils/parsers/string/AtomOrStringParser.java

```diff
@@ -76,4 +81,52 @@ private void append(AppendableCharSequence seq, char c) {
     seq.append(c);
   }
 
+  /**
+   * See https://tools.ietf.org/html/rfc3501.html#section-4.3
+   *
+   * String literals have the form
+   *
+   * {10}
+   * abcdefghij
+   *
+   * Where {10} represents the length of the string literal. The literal
+   * begins after any CLRF characters following the '}' character.
+   */
+  private String parseLiteral(ByteBuf buffer, AppendableCharSequence seq) {
+    int length = 0;
+
+    char c = (char) buffer.readUnsignedByte();
+    while (c != '}') {
+      length = (length * 10) + Character.getNumericValue(c);
+      c = (char) buffer.readUnsignedByte();
+    }
+
+    if (length > 0) {
+      c = (char) buffer.readUnsignedByte();
+      while (isCLRFCharacter(c)) {
+        c = (char) buffer.readUnsignedByte();
+      }
+
+      seq.append(c);
+      length -= 1;
+      while (length > 0) {
+        seq.append((char) buffer.readUnsignedByte());
+        length--;
+      }
+
+      if (buffer.readableBytes() != 0) {
+        String originalBuffer = seq.toString()
+            + buffer.toString(StandardCharsets.UTF_8);
+        throw new DecoderException("String literal doesn't match expected length " + originalBuffer);
```

Comment:
I think originalBuffer will almost always have leftover bytes after you are done reading the literal sequence, so this will throw more than you want it to.